File: backend/app/models/planner.py
"""Mood Planner + Event Planner API models (spec §14, §15)."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class MoodPlanRequest(BaseModel):
    """ "How do you want to feel today?" — and optionally, what for.

    Deliberately has NO render field. Escalating a plan into a paid render is a
    separate, explicit call to POST /v1/tryon (§14).
    """

    mood: str = Field(max_length=40)
    occasion: str | None = Field(default=None, max_length=40)


class MoodPlanResponse(BaseModel):
    id: str
    mood: str
    occasion: str | None = None
    headline: str
    lines: list[str] = Field(default_factory=list)
    #: Wardrobe item ids the direction referenced, in build order. Empty when
    #: the closet is too thin to name real pieces.
    item_ids: list[str] = Field(default_factory=list)
    #: True when no real pieces could be named — the app must not claim the
    #: direction was picked from the user's closet.
    generic: bool = False
    created_at: datetime


class StyleEventIn(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    event_at: datetime
    occasion: str | None = Field(default=None, max_length=40)
    #: The saved look this event is dressed for. Saved looks are device-local,
    #: so this is the look's stable id plus its durable image URL.
    look_ref: str | None = Field(default=None, max_length=120)
    look_image_url: str | None = Field(default=None, max_length=2000)
    note: str | None = Field(default=None, max_length=500)
    #: Opt-in per event. Default false: creating an event must never sign a user
    #: up for push (§23).
    reminder_opt_in: bool = False


class StyleEventUpdate(BaseModel):
    """Every field optional — a PATCH only changes what it names."""

    name: str | None = Field(default=None, min_length=1, max_length=120)
    event_at: datetime | None = None
    occasion: str | None = Field(default=None, max_length=40)
    look_ref: str | None = Field(default=None, max_length=120)
    look_image_url: str | None = Field(default=None, max_length=2000)
    note: str | None = Field(default=None, max_length=500)
    reminder_opt_in: bool | None = None


class StyleEvent(BaseModel):
    id: str
    name: str
    event_at: datetime
    occasion: str | None = None
    look_ref: str | None = None
    look_image_url: str | None = None
    note: str | None = None
    reminder_opt_in: bool = False
    created_at: datetime


class StyleEventList(BaseModel):
    events: list[StyleEvent] = Field(default_factory=list)
    #: The soonest upcoming event, repeated here so Home does not have to
    #: re-derive it from a list it may have paged.
    next_event: StyleEvent | None = None

File: app/lib/data/models/planner.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'planner.freezed.dart';
part 'planner.g.dart';

/// The four mood zones the board already uses (`ui/home/wtm_mood.dart`).
///
/// The planner speaks the SAME vocabulary as the existing mood slider rather
/// than inventing a second one, so a mood set on Home and a mood picked in the
/// planner are the same fact.
enum PlannerMood {
  calm('calm'),
  confident('confident'),
  bold('bold'),
  rebel('rebel');

  const PlannerMood(this.wire);

  final String wire;

  /// The zone for a 0..1 slider value — the same four-way split as
  /// [WtmMoodZone], so the two can never disagree about where `bold` starts.
  static PlannerMood ofValue(double value) => switch (value) {
    < 0.25 => calm,
    < 0.5 => confident,
    < 0.75 => bold,
    _ => rebel,
  };
}

/// Optional context for a plan. Null is a valid answer: the user may simply
/// have picked a feeling.
enum PlannerOccasion {
  everyday('everyday'),
  work('work'),
  date('date'),
  brunch('brunch'),
  wedding('wedding'),
  nightOut('night_out');

  const PlannerOccasion(this.wire);

  final String wire;
}

/// Styling direction for a mood. **Costs nothing and renders nothing** —
/// turning a plan into an image is a separate, explicit, paid action.
@freezed
abstract class MoodPlan with _$MoodPlan {
  const factory MoodPlan({
    required String id,
    required String mood,
    String? occasion,
    @Default('') String headline,
    @Default(<String>[]) List<String> lines,

    /// Wardrobe item ids the direction referenced, in build order.
    @JsonKey(name: 'item_ids') @Default(<String>[]) List<String> itemIds,

    /// True when no real pieces could be named. The UI must NOT claim the
    /// direction was picked from the user's closet in that case.
    @Default(false) bool generic,
    @JsonKey(name: 'created_at') required DateTime createdAt,
  }) = _MoodPlan;

  factory MoodPlan.fromJson(Map<String, dynamic> json) =>
      _$MoodPlanFromJson(json);
}

/// A user-entered event. No calendar is read or written (spec §15).
@freezed
abstract class StyleEvent with _$StyleEvent {
  const factory StyleEvent({
    required String id,
    required String name,
    @JsonKey(name: 'event_at') required DateTime eventAt,
    String? occasion,

    /// The saved look this event is dressed for. Saved looks live on the
    /// device, so this is the look's stable id plus its durable image URL.
    @JsonKey(name: 'look_ref') String? lookRef,
    @JsonKey(name: 'look_image_url') String? lookImageUrl,
    String? note,

    /// Opt-in per event. Creating an event never signs the user up for push.
    @JsonKey(name: 'reminder_opt_in') @Default(false) bool reminderOptIn,
    @JsonKey(name: 'created_at') DateTime? createdAt,
  }) = _StyleEvent;

  const StyleEvent._();

  factory StyleEvent.fromJson(Map<String, dynamic> json) =>
      _$StyleEventFromJson(json);

  /// Whole days until the event, floor-clamped at zero. Used for the "in 3
  /// days" line — computed from the device clock on purpose, because that is
  /// the clock the user is reading the screen on.
  int get daysAway {
    final diff = eventAt.difference(DateTime.now());
    return diff.isNegative ? 0 : diff.inDays;
  }

  bool get isPast => eventAt.isBefore(DateTime.now());
}

@freezed
abstract class StyleEventList with _$StyleEventList {
  const factory StyleEventList({
    @Default(<StyleEvent>[]) List<StyleEvent> events,
    @JsonKey(name: 'next_event') StyleEvent? nextEvent,
  }) = _StyleEventList;

  factory StyleEventList.fromJson(Map<String, dynamic> json) =>
      _$StyleEventListFromJson(json);
}
